### PR TITLE
Library search slowing down a lot after putting space while typing in search bar

### PR DESCRIPTION
**Pull request by vmoyseenko**
#### Purpose

Search is too slow for phrases which contains spaces. This PR tries to fix it.

#### Modifications

There was investigated that main reason of slowness is using of Regular Expressions. It is proposed to replace regualar expression by manual implementation. Because of used regular expression manual implementation looks not so complicated.

#### Performance check

All presented data counted in milliseconds. Were done 3 passes. Measured just search algorithm. Not the time needed for visual population of search results.

Letter entered|Old impl. average|New impl. average
--------------|-----------------|-----------------
d|	1.000|3.667
o|	0.667|1.000
u|	1.000|1.333
b|	1.000|1.000
l|	1.000|1.333
e|	1.000|1.000
 |	0.667|1.000
s|	523.697|1.333
l|	582.367|1.333
i|	604.368|1.333
d|	593.034|1.000
e|	591.034|1.000
r|	588.034|1.000

Current computer specification: Intel i7, 8GB RAM.

By the way as I noticed for regular expressions on i5 `slider` times are near 930 ms.

Let me know if more search phrases should be measured.

#### Search content
Here is below a couple snapshots of content for old implementation (left) and new implementation (right).
capture

Below found a case were results doesn't match. Which of sides is correct?
capture

For next letter results are the same.
capture

capture

capture

#### Additional references

[MAGN-6018](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6018) Library search slowing down a lot after putting space while typing in search bar.

#### Unit tests
capture

#### Reviewers

@pboyer, @Benglin, please take a look.

### CODE REVIEW

File: src/DynamoCoreWpf/Views/Search/LibrarySearchView.xaml [hunk 1]
```diff
@@ -710,7 +710,7 @@
                                                     </TextBlock>
                                                     <ListBox Background="Transparent"
                                                              BorderBrush="Transparent"
-                                                             ItemContainerStyle="{DynamicResource ListBoxMembersStyle}"
+                                                             ItemContainerStyle="{StaticResource ListBoxMembersStyle}"
```
Comment (vmoyseenko) [inline src/DynamoCoreWpf/Views/Search/LibrarySearchView.xaml]: Binding errors in output. Fixed them here.

**Comment by vmoyseenko** [comment]
@pboyer, @Benglin, please, take a look. We have difference in search results. You can find it on screenshot for `double s`. How should I handle it?

File: src/DynamoCoreWpf/ViewModels/Search/SearchCategory.cs [hunk 2]
```diff
@@ -8,9 +9,7 @@ namespace Dynamo.Search
 {
     public class SearchCategory
     {
-
-        // TODO: classes functionality.
-        //private readonly ObservableCollection<NodeCategoryViewModel> classes;
+        private readonly ObservableCollection<NodeCategoryViewModel> classes;
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/Search/SearchCategory.cs]: Do we want to enable classes now? If we do, can it not be part of another PR?

File: src/DynamoCoreWpf/Views/Search/LibrarySearchView.xaml (hunk 1 above)
Comment (Benglin) [inline src/DynamoCoreWpf/Views/Search/LibrarySearchView.xaml]: If this fixes the output in debugger, then perhaps it can be made a standalone PR so it can be quickly merged in. Also, @ke-yu has reported another exception when we start debugging Dynamo (with .NET exceptions enabled), if you could fix this also in the same standalone PR, that will be really helpful:

    "Object reference not set to an instance of an object."
    at Controls\controls\HeaderStrip.xaml:line 85

**Comment by Benglin** [comment]
Thanks for the fix, @vmoyseenko! This surely is exciting, look at that constant speed search! :)

As for the different result list when you searched for `double`, I think none of the two lists is *right*. Do you think we should not search into the argument list? Maybe @elayabharath can advise what's the desirable behavior here, but I think searching the argument string is a little overkill.

**Comment by pboyer** [comment]
Hi @vmoyseenko, 

This is a clear improvement.  I think there's opportunities to improve this algorithm a bit (e.g. search by characters and ignoring white space) and we may use it as the only search algorithm, but for now this fixes a critical performance issue.  

I differ with @Benglin  on argument lists.  I thin it's fine as it is.

I'm going to merge this in as it is.  If we want to make any of the changes @Benglin has made, we can tackle them in subsequent PR's.  

Thanks!

File: src/DynamoCoreWpf/ViewModels/Search/SearchCategory.cs (hunk 2 above)
Comment (vmoyseenko) [inline src/DynamoCoreWpf/ViewModels/Search/SearchCategory.cs]: We need it to fix Binding errors.
